README.md
# [keyfetch](https://git.rootprojects.org/root/keyfetch.js)

Lightweight support for fetching JWKs.

Fetches JSON native JWKs and exposes them as PEMs that can be consumed by the `jsonwebtoken` package
(and node's native RSA and ECDSA crypto APIs).

## Features

Works great for

-   [x] `jsonwebtoken` (Auth0)
-   [x] OIDC (OpenID Connect)
-   [x] .well-known/jwks.json (Auth0, Okta)
-   [x] Other JWKs URLs

Crypto Support

-   [x] JWT verification
-   [x] RSA (all variants)
-   [x] EC / ECDSA (NIST variants P-256, P-384)
-   [x] Sane error codes
-   [ ] esoteric variants (excluded to keep the code featherweight and secure)

# Table of Contents

-   Install
-   Usage
-   API
    -   Auth0 / Okta
    -   OIDC
-   Errors
-   Change Log

# Install

```bash
npm install --save keyfetch
```

# Usage

Retrieve a key list of keys:

```js
var keyfetch = require("keyfetch");

keyfetch.oidcJwks("https://example.com/").then(function (results) {
    results.forEach(function (result) {
        console.log(result.jwk);
        console.log(result.thumprint);
        console.log(result.pem);
    });
});
```

Quick JWT verification (for authentication):

```js
var keyfetch = require("keyfetch");
var jwt = "...";

keyfetch.jwt.verify(jwt).then(function (decoded) {
    console.log(decoded);
});
```

JWT verification (for authorization):

```js
var options = { issuers: ["https://example.com/"], claims: { role: "admin" } };
keyfetch.jwt.verify(jwt, options).then(function (decoded) {
    console.log(decoded);
});
```

Verify a JWT with `jsonwebtoken`:

```js
var keyfetch = require("keyfetch");
var jwt = require("jsonwebtoken");
var auth = "..."; // some JWT
var token = jwt.decode(auth, { json: true, complete: true });

if (!isTrustedIssuer(token.payload.iss)) {
    throw new Error("untrusted issuer");
}

keyfetch.oidcJwk(token.header.kid, token.payload.iss).then(function (result) {
    console.log(result.jwk);
    console.log(result.thumprint);
    console.log(result.pem);

    jwt.jwt.verify(jwt, { jwk: result.jwk });
});
```

_Note_: You might implement `isTrustedIssuer` one of these:

```js
function isTrustedIssuer(iss) {
    return -1 !== ["https://partner.com/", "https://auth0.com/"].indexOf(iss);
}
```

```js
function isTrustedIssuer(iss) {
    return (
        /^https:/.test(iss) && /(\.|^)example\.com$/.test(iss) // must be a secure domain
    ); // can be example.com or any subdomain
}
```

# API

All API calls will return the RFC standard JWK SHA256 thumbprint as well as a PEM version of the key.

Note: When specifying `id`, it may be either `kid` (as in `token.header.kid`)
or `thumbprint` (as in `result.thumbprint`).

### JWKs URLs

Retrieves keys from a URL such as `https://example.com/jwks/` with the format `{ keys: [ { kid, kty, exp, ... } ] }`
and returns the array of keys (as well as thumbprint and jwk-to-pem).

```js
keyfetch.jwks(jwksUrl);
// Promises [ { jwk, thumbprint, pem } ] or fails
```

```js
keyfetch.jwk(id, jwksUrl);
// Promises { jwk, thumbprint, pem } or fails
```

### Auth0

If `https://example.com/` is used as `issuerUrl` it will resolve to
`https://example.com/.well-known/jwks.json` and return the keys.

```js
keyfetch.wellKnownJwks(issuerUrl);
// Promises [ { jwk, thumbprint, pem } ] or fails
```

```js
keyfetch.wellKnownJwk(id, issuerUrl);
// Promises { jwk, thumbprint, pem } or fails
```

### OIDC

If `https://example.com/` is used as `issuerUrl` then it will first resolve to
`https://example.com/.well-known/openid-configuration` and then follow `jwks_uri` to return the keys.

```js
keyfetch.oidcJwks(issuerUrl);
// Promises [ { jwk, thumbprint, pem } ] or fails
```

```js
keyfetch.oidcJwk(id, issuerUrl);
// Promises { jwk, thumbprint, pem } or fails
```

### Verify JWT

This can accept a _JWT string_ (compact JWS) or a _decoded JWT object_ (JWS).

This can be used purely for verifying pure authentication tokens, as well as authorization tokens.

```js
keyfetch.jwt.verify(jwt, { strategy: "oidc" }).then(function (verified) {
    /*
    { protected: '...'  // base64 header
    , payload: '...'    // base64 payload
    , signature: '...'  // base64 signature
    , header: {...}     // decoded header
    , claims: {...}     // decoded payload
    }
  */
});
```

When used for authorization, it's important to specify a limited set of trusted `issuers`. \
When using for federated authentication you may set `issuers = ["*"]` - but **DO NOT** trust claims such as `email` and `email_verified`.

If your authorization `claims` can be expressed as exact string matches, you can specify those too.

```js
keyfetch.jwt.verify(jwt, {
  strategy: 'oidc',
  issuers: [ 'https://example.com/' ],
  //iss: 'https://example.com/',
  claims: { role: 'admin', sub: 'abc', group: 'xyz' }
}).then(function (verified) {
```

-   `strategy` may be `oidc` (default) , `auth0`, or a direct JWKs url.
-   `issuers` must be a list of https urls (though http is allowed for things like Docker swarm), or '\*'
-   `iss` is like `issuers`, but only one
-   `claims` is an object with arbitrary keys (i.e. everything except for the standard `iat`, `exp`, `jti`, etc)
-   `exp` may be set to `false` if you're validating on your own (i.e. allowing time drift leeway)
-   `jwks` can be used to specify a list of allowed public key rather than fetching them (i.e. for offline unit tests)
-   `jwk` same as above, but a single key rather than a list

### Decode JWT

```jwt
try {
  console.log( keyfetch.jwt.decode(jwt) );
} catch(e) {
  console.error(e);
}
```

```js
{ protected: '...'  // base64 header
, payload: '...'    // base64 payload
, signature: '...'  // base64 signature
, header: {...}     // decoded header
, claims: {...}     // decoded payload
```

It's easier just to show the code than to explain the example.

```js
keyfetch.jwt.decode = function (jwt) {
    // Unpack JWS from "compact" form
    var parts = jwt.split(".");
    var obj = {
        protected: parts[0],
        payload: parts[1],
        signature: parts[2]
    };

    // Decode JWT properties from JWS as unordered objects
    obj.header = JSON.parse(Buffer.from(obj.protected, "base64"));
    obj.claims = JSON.parse(Buffer.from(obj.payload, "base64"));

    return obj;
};
```

### Cache Settings

```js
keyfetch.init({
    // set all keys at least 1 hour (regardless of jwk.exp)
    mincache: 1 * 60 * 60,

    // expire each key after 3 days (regardless of jwk.exp)
    maxcache: 3 * 24 * 60 * 60,

    // re-fetch a key up to 15 minutes before it expires (only if used)
    staletime: 15 * 60
});
```

There is no background task to cleanup expired keys as of yet.
For now you can limit the number of keys fetched by having a simple whitelist.

# Errors

`JSON.stringify()`d errors look like this:

```js
{
  code: "INVALID_JWT",
  status: 401,
  details: [ "jwt.claims.exp = 1634804500", "DEBUG: helpful message" ]
  message: "token's 'exp' has passed or could not parsed: 1634804500"
}
```

SemVer Compatibility:

-   `code` & `status` will remain the same.
-   `message` is **NOT** included in the semver compatibility guarantee (we intend to make them more client-friendly), neither is `detail` at this time (but it will be once we decide on what it should be).
-   `details` may be added to, but not subtracted from

| Hint              | Code          | Status | Message (truncated)                                    |
| ----------------- | ------------- | ------ | ------------------------------------------------------ |
| bad gateway       | BAD_GATEWAY   | 502    | The auth token could not be verified because our se... |
| insecure issuer   | MALFORMED_JWT | 400    | The auth token could not be verified because our se... |
| parse error       | MALFORMED_JWT | 400    | The auth token could not be verified because it is ... |
| no issuer         | MALFORMED_JWT | 400    | The auth token could not be verified because it doe... |
| malformed exp     | MALFORMED_JWT | 400    | The auth token could not be verified because it's e... |
| expired           | INVALID_JWT   | 401    | The auth token is expired. To try again, go to the ... |
| inactive          | INVALID_JWT   | 401    | The auth token isn't valid yet. It's activation dat... |
| bad signature     | INVALID_JWT   | 401    | The auth token did not pass verification because it... |
| jwk not found old | INVALID_JWT   | 401    | The auth token did not pass verification because ou... |
| jwk not found     | INVALID_JWT   | 401    | The auth token did not pass verification because ou... |
| no jwkws uri      | INVALID_JWT   | 401    | The auth token did not pass verification because it... |
| unknown issuer    | INVALID_JWT   | 401    | The auth token did not pass verification because it... |
| failed claims     | INVALID_JWT   | 401    | The auth token did not pass verification because it... |

# Change Log

Minor Breaking changes (with a major version bump):

-   v3.0.0
    -   reworked error messages (also available in v2.1.0 as `client_message`)
    -   started using `let` and template strings (drops _really_ old node compat)
-   v2.0.0
    -   changes from the default `issuers = ["*"]` to requiring that an issuer (or public jwk for verification) is specified

See other changes in [CHANGELOG.md](./CHANGELOG.md).


CHANGELOG.md
# v3.0.0

**Breaking Change**: Standardize error `message`s (now they're more client-friendly).

# v2.1.0

Feature: Add `code`, `status`, and `details` to errors.

# v2.0.0

**Breaking Change**: require `issuers` array (rather than `["*"]` by default).


keyfetch-test.js
"use strict";

var keyfetch = require("./keyfetch.js");
var testIss = "https://example.auth0.com";

keyfetch.init({});
keyfetch
    .oidcJwks(testIss)
    .then(function (hits) {
        keyfetch._clear();
        //console.log(hits);
        return keyfetch.oidcJwk(hits[0].thumbprint, testIss).then(function () {
            return keyfetch.oidcJwk(hits[0].thumbprint, testIss).then(function (/*jwk*/) {
                //console.log(jwk);
            });
        });
    })
    .then(function () {
        console.log("Fetching PASSES");
    })
    .catch(function (err) {
        console.error("NONE SHALL PASS!");
        console.error(err);
        process.exit(1);
    });

/*global Promise*/
var keypairs = require("keypairs");
keypairs.generate().then(function (pair) {
    var iss = "https://example.com/";
    return Promise.all([
        keypairs
            .signJwt({
                jwk: pair.private,
                iss: iss,
                sub: "mikey",
                exp: "1h"
            })
            .then(function (jwt) {
                return Promise.all([
                    keyfetch.jwt.verify(jwt, { jwk: pair.public, iss: "*" }).then(function (verified) {
                        if (!(verified.claims && verified.claims.exp)) {
                            throw new Error("malformed decoded token");
                        }
                    }),
                    keyfetch.jwt
                        .verify(keyfetch.jwt.decode(jwt), { jwk: pair.public, iss: iss })
                        .then(function (verified) {
                            if (!(verified.claims && verified.claims.exp)) {
                                throw new Error("malformed decoded token");
                            }
                        }),
                    keyfetch.jwt.verify(jwt, { jwks: [pair.public], issuers: [iss] }),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        issuers: ["https://example.com/"]
                    }),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        issuers: ["https://example.com"]
                    }),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        issuers: ["example.com"]
                    }),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        issuers: ["example.com/"]
                    }),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        issuers: ["*"]
                    }),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            issuers: ["http://example.com"]
                        })
                        .then(e("bad scheme"))
                        .catch(throwIfNotExpected),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            issuers: ["https://www.example.com"]
                        })
                        .then(e("bad prefix"))
                        .catch(throwIfNotExpected),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            issuers: ["https://wexample.com"]
                        })
                        .then(e("bad sld"))
                        .catch(throwIfNotExpected),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            issuers: ["https://example.comm"]
                        })
                        .then(e("bad tld"))
                        .catch(throwIfNotExpected),
                    keyfetch.jwt.verify(jwt, {
                        jwk: pair.public,
                        claims: { iss: "https://example.com/" }
                    }),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            claims: { iss: "https://example.com" }
                        })
                        .then(e("inexact claim"))
                        .catch(throwIfNotExpected)
                ]);
            }),
        keypairs
            .signJwt({
                jwk: pair.private,
                iss: false,
                sub: "mikey",
                exp: "1h"
            })
            .then(function (jwt) {
                return Promise.all([
                    // test that the old behavior of defaulting to '*' still works
                    keyfetch.jwt
                        .verify(jwt, { jwk: pair.public })
                        .then(e("should have issued security warning about allow all by default"))
                        .catch(throwIfNotExpected),
                    keyfetch.jwt.verify(jwt, { jwk: pair.public, issuers: ["*"] }),
                    keyfetch.jwt.verify(jwt).then(e("should have an issuer")).catch(throwIfNotExpected),
                    keyfetch.jwt
                        .verify(jwt, {
                            jwk: pair.public,
                            issuers: ["https://example.com/"]
                        })
                        .then(e("fail when issuer specified and doesn't exist"))
                        .catch(throwIfNotExpected)
                ]);
            })
    ])
        .then(function () {
            console.log("JWT PASSES");
        })
        .catch(function (err) {
            console.error("NONE SHALL PASS!");
            console.error(err);
            process.exit(1);
        });
});
/*
var jwt = '...';
keyfetch.verify({ jwt: jwt }).catch(function (err) {
  console.log(err);
});
*/

function e(msg) {
    return new Error("ETEST: " + msg);
}
function throwIfNotExpected(err) {
    if ("ETEST" === err.message.slice(0, 5)) {
        throw err;
    }
}


lib/errors.js
"use strict";

// Possible User Errors

/**
 * @typedef AuthError
 * @property {string} message
 * @property {number} status
 * @property {string} code
 * @property {any} [details]
 */

/**
 * @param {string} msg
 * @param {{
 *   status: number,
 *   code: string,
 *   details?: any,
 * }} opts
 * @returns {AuthError}
 */
function create(old, msg, code, status, details) {
    /** @type AuthError */
    //@ts-ignore
    let err = new Error(msg);
    err.message = msg;
    err._old_message = old;
    err.code = code;
    err.status = status;
    if (details) {
        err.details = details;
    }
    err.source = "keyfetch";
    err.toJSON = toJSON;
    err.toString = toString;
    return err;
}

function toJSON() {
    /*jshint validthis:true*/
    return {
        message: this.message,
        status: this.status,
        code: this.code,
        details: this.details
    };
}
function toString() {
    /*jshint validthis:true*/
    return this.stack + "\n" + JSON.stringify(this);
}

// DEVELOPER_ERROR - a good token won't make a difference
var E_DEVELOPER = "DEVELOPER_ERROR";

// BAD_GATEWAY - there may be a temporary error fetching the public or or whatever
var E_BAD_GATEWAY = "BAD_GATEWAY";

// MALFORMED_JWT - the token could not be verified - not parsable, missing claims, etc
var E_MALFORMED = "MALFORMED_JWT";

// INVALID_JWT - the token's properties don't meet requirements - iss, claims, sig, exp
var E_INVALID = "INVALID_JWT";

module.exports = {
    //
    // DEVELOPER_ERROR (dev / server)
    //

    /**
     * @param {string} msg
     * @returns {AuthError}
     */
    DEVELOPER_ERROR: function (old, msg, details) {
        return create(old, msg || old, E_DEVELOPER, 500, details);
    },
    BAD_GATEWAY: function (err) {
        var msg =
            "The auth token could not be verified because our server encountered a network error (or a bad gateway) when connecting to its issuing server.";
        var details = [];
        if (err.message) {
            details.push("error.message = " + err.message);
        }
        if (err.response && err.response.statusCode) {
            details.push("response.statusCode = " + err.response.statusCode);
        }
        return create(msg, msg, E_BAD_GATEWAY, 502, details);
    },

    //
    // MALFORMED_TOKEN (dev / client)
    //

    /**
     * @param {string} iss
     * @returns {AuthError}
     */
    INSECURE_ISSUER: function (iss) {
        var old =
            "'" + iss + "' is NOT secure. Set env 'KEYFETCH_ALLOW_INSECURE_HTTP=true' to allow for testing. (iss)";
        var details = [
            "jwt.claims.iss = " + JSON.stringify(iss),
            "DEBUG: Set ENV 'KEYFETCH_ALLOW_INSECURE_HTTP=true' to allow insecure issuers (for testing)."
        ];
        var msg =
            'The auth token could not be verified because our server could connect to its issuing server ("iss") securely.';
        return create(old, msg, E_MALFORMED, 400, details);
    },
    /**
     * @param {string} jwt
     * @returns {AuthError}
     */
    PARSE_ERROR: function (jwt) {
        var old = "could not parse jwt: '" + jwt + "'";
        var msg = "The auth token could not be verified because it is malformed.";
        var details = ["jwt = " + JSON.stringify(jwt)];
        return create(old, msg, E_MALFORMED, 400, details);
    },
    /**
     * @param {string} iss
     * @returns {AuthError}
     */
    NO_ISSUER: function (iss) {
        var old = "'iss' is not defined";
        var msg = 'The auth token could not be verified because it doesn\'t specify an issuer ("iss").';
        var details = ["jwt.claims.iss = " + JSON.stringify(iss)];
        return create(old, msg, E_MALFORMED, 400, details);
    },

    /**
     * @param {string} iss
     * @returns {AuthError}
     */
    MALFORMED_EXP: function (exp) {
        var old = "token's 'exp' has passed or could not parsed: '" + exp + "'";
        var msg = 'The auth token could not be verified because it\'s expiration date ("exp") could not be read';
        var details = ["jwt.claims.exp = " + JSON.stringify(exp)];
        return create(old, msg, E_MALFORMED, 400, details);
    },

    //
    // INVALID_TOKEN (dev / client)
    //

    /**
     * @param {number} exp
     * @returns {AuthError}
     */
    EXPIRED: function (exp) {
        var old = "token's 'exp' has passed or could not parsed: '" + exp + "'";
        // var msg = "The auth token did not pass verification because it is expired.not properly signed.";
        var msg = "The auth token is expired. To try again, go to the main page and sign in.";
        var details = ["jwt.claims.exp = " + JSON.stringify(exp)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /**
     * @param {number} nbf
     * @returns {AuthError}
     */
    INACTIVE: function (nbf) {
        var old = "token's 'nbf' has not been reached or could not parsed: '" + nbf + "'";
        var msg = "The auth token isn't valid yet. It's activation date (\"nbf\") is in the future.";
        var details = ["jwt.claims.nbf = " + JSON.stringify(nbf)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /** @returns {AuthError} */
    BAD_SIGNATURE: function (jwt) {
        var old = "token signature verification was unsuccessful";
        var msg = "The auth token did not pass verification because it is not properly signed.";
        var details = ["jwt = " + JSON.stringify(jwt)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /**
     * @param {string} kid
     * @returns {AuthError}
     */
    JWK_NOT_FOUND_OLD: function (kid) {
        var old = "Retrieved a list of keys, but none of them matched the 'kid' (key id) of the token.";
        var msg =
            'The auth token did not pass verification because our server couldn\'t find a mutually trusted verification key ("jwk").';
        var details = ["jws.header.kid = " + JSON.stringify(kid)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /**
     * @param {string} id
     * @returns {AuthError}
     */
    JWK_NOT_FOUND: function (id) {
        // TODO Distinguish between when it's a kid vs thumbprint.
        var old = "No JWK found by kid or thumbprint '" + id + "'";
        var msg =
            'The auth token did not pass verification because our server couldn\'t find a mutually trusted verification key ("jwk").';
        var details = ["jws.header.kid = " + JSON.stringify(id)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /** @returns {AuthError} */
    NO_JWKWS_URI: function (url) {
        var old = "Failed to retrieve openid configuration";
        var msg =
            'The auth token did not pass verification because its issuing server did not list any verification keys ("jwks").';
        var details = ["OpenID Provider Configuration: " + JSON.stringify(url)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /**
     * @param {string} iss
     * @returns {AuthError}
     */
    UNKNOWN_ISSUER: function (iss) {
        var old = "token was issued by an untrusted issuer: '" + iss + "'";
        var msg = "The auth token did not pass verification because it wasn't issued by a server that we trust.";
        var details = ["jwt.claims.iss = " + JSON.stringify(iss)];
        return create(old, msg, E_INVALID, 401, details);
    },
    /**
     * @param {Array<string>} details
     * @returns {AuthError}
     */
    FAILED_CLAIMS: function (details, claimNames) {
        var old = "token did not match on one or more authorization claims: '" + claimNames + "'";
        var msg =
            'The auth token did not pass verification because it failed some of the verification criteria ("claims").';
        return create(old, msg, E_INVALID, 401, details);
    }
};
var Errors = module.exports;

// for README
if (require.main === module) {
    let maxWidth = 54;
    let header = ["Hint", "Code", "Status", "Message (truncated)"];
    let widths = header.map(function (v) {
        return Math.min(maxWidth, String(v).length);
    });
    let rows = [];
    Object.keys(module.exports).forEach(function (k) {
        //@ts-ignore
        var E = module.exports[k];
        var e = E("test");
        var code = e.code;
        var msg = e.message;
        var hint = k.toLowerCase().replace(/_/g, " ");
        widths[0] = Math.max(widths[0], String(hint).length);
        widths[1] = Math.max(widths[1], String(code).length);
        widths[2] = Math.max(widths[2], String(e.status).length);
        widths[3] = Math.min(maxWidth, Math.max(widths[3], String(msg).length));
        rows.push([hint, code, e.status, msg]);
    });
    rows.forEach(function (cols, i) {
        let cells = cols.map(function (col, i) {
            if (col.length > maxWidth) {
                col = col.slice(0, maxWidth - 3);
                col += "...";
            }
            return String(col).padEnd(widths[i], " ");
        });
        let out = `| ${cells[0]} | ${cells[1]} | ${cells[2]} | ${cells[3].slice(0, widths[3])} |`;
        //out = out.replace(/\| /g, " ").replace(/\|/g, "");
        console.info(out);
        if (i === 0) {
            cells = cols.map(function (col, i) {
                return "-".padEnd(widths[i], "-");
            });
            console.info(`| ${cells[0]} | ${cells[1]} | ${cells[2]} | ${cells[3]} |`);
        }
    });
    console.log();
    console.log(Errors.MALFORMED_EXP());
    console.log();
    console.log(JSON.stringify(Errors.MALFORMED_EXP(), null, 2));
}
